tests/test_fips.py

```diff
@@ -0,0 +1,215 @@
+"""This module checks whether the container images run in FIPS mode on a host in
+FIPS mode.
+
+"""
+import sys
+
+import pytest
+from _pytest.config import Config
+from _pytest.mark.structures import ParameterSet
+from bci_tester.data import ALL_CONTAINERS
+from bci_tester.data import BASE_CONTAINER
+from bci_tester.data import BUSYBOX_CONTAINER
+from bci_tester.data import CONTAINERS_WITH_ZYPPER
+from bci_tester.data import CONTAINERS_WITHOUT_ZYPPER
+from bci_tester.data import OS_VERSION
+from bci_tester.fips import FIPS_DIGESTS
+from bci_tester.fips import host_fips_enabled
+from bci_tester.fips import NONFIPS_DIGESTS
+from pytest_container.build import MultiStageBuild
+from pytest_container.helpers import get_extra_build_args
+from pytest_container.helpers import get_extra_run_args
+from pytest_container.runtime import OciRuntimeBase
+
+
+# building the documentation will fail on a non-FIPS host otherwise
+if "sphinx" not in sys.modules:
+    assert (
+        host_fips_enabled()
+    ), "The host must run in FIPS mode for the FIPS test suite"
+
+
+#: Error message from OpenSSL when a non-FIPS digest is selected in FIPS mode
+FIPS_ERR_MSG = (
+    "not a known digest" if OS_VERSION == "15.3" else "Error setting digest"
+)
+
+
+#: multistage :file:`Dockerfile` that builds the program from
+#: :py:const:`FIPS_TEST_DOT_C` using gcc and copies it, ``libcrypto``, ``libssl``
+#: and ``libz`` into the deployment image. The libraries must be copied, as they
+#: are not available in the minimal container images.
+DOCKERFILE = """FROM $builder as builder
+
+WORKDIR /src/
+COPY fips-test.c /src/
+RUN zypper -n ref && zypper -n in gcc libopenssl-devel && zypper -n clean
+RUN gcc -Og -g3 fips-test.c -Wall -Wextra -Wpedantic -lcrypto -lssl -o fips-test
+
+FROM $runner
+
+COPY --from=builder /src/fips-test /bin/fips-test
+COPY --from=builder /usr/lib64/libcrypto.so.1.1 /usr/lib64/
+COPY --from=builder /usr/lib64/libssl.so.1.1 /usr/lib64/
+COPY --from=builder /lib64/libz.so.1 /usr/lib64/
+RUN /bin/fips-test sha256
+"""
+
+#: Simple C program that will calculate the hash of a fixed string using openSSL.
+#: The hash to be used is passed as the first parameter to the binary.
+#: This program has been taken from
+#: `<https://www.openssl.org/docs/manmaster/man3/EVP_DigestInit.html>`_
+FIPS_TEST_DOT_C = r"""#include <openssl/evp.h>
```

Comment:
Is there no way of putting this in an external file? Working on inline code is somehow cumbersome.

Reply:
There is, it's now under `tests/files/`